[Bugfix] handle prompt_logprobs in _apply_min_tokens_penalty

tjohnson31415 opened the pull request:
I implemented `_apply_min_tokens_penalty` in https://github.com/vllm-project/vllm/pull/3124, but was not aware that having `prompt_logprobs` in the sampling parameters affects the shape of `logits`. This PR fixes a bug that would occur if using `min_tokens` and `prompt_logprobs` together where the `start_idx` cursor would not be tracking across sequences correctly. The fix is to detect sequences with `prompt_logprobs` and skip over the rows in logits for all but the last token in the prompt.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

tjohnson31415 committed:
[bugfix] handle prompt_logprobs in _apply_min_tokens_penalty

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

njhill reviewed (approved):
Thanks @tjohnson31415!

Do you think you could add a test? Could just set `stop_token_ids` to be the id of the first token expected?

tjohnson31415 committed:
doc: add comment about logits with prompt_logprobs

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
test: add prompt_logprobs tests for apply_min_tokens_penalty

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 commented:
@njhill I have added some tests with `prompt_logprobs`: I added a few more test cases to the `test_sampler_min_tokens_penalty` tests and updated the request generation / validation accordingly.

I also added a few sentences to the Sampler doc string to describe the situation with `prompt_logprobs` adding rows to logits. Let me know if that is clear.

njhill commented:
@tjohnson31415 this looks great thank you! Sorry, I didn't realize the test changes would be so involved.

This changes the tests to assume that each batch will comprise only either all prompt or all decode seq groups right? Which will always be the case right now but I'm concerned might not be soon with the [chunked prefill](https://github.com/vllm-project/vllm/issues/3130) changes. So we may need to change the test(s) again at that point to cover hybrid batches.

FYI @rkooo567 ... with hybrid batches, will ordering be such that all prefill sequences will come before the decode sequences?

rkooo567 commented:
@njhill yeah ordered by prefill -> decode (within input_tokens).  The ordering can be controlled from the scheduler.
